### PR TITLE
removed last of pkg_resources.

### PR DESCRIPTION
**Pull request by ChrisBarker-NOAA**
I've removed the last of pkg_resources, replacing it with:

`importlib.metadata`

for Python >= 3.9 

and the backport:

`import importlib_metadata`

for older versions.

I also added conditional import of: `importlib.resources`, so the backport will only be used if it's needed.

Both the backports are in the requirements files, as I know of no way to make a requirement python-version-specific.

NOTE: do we really need to keep older versions working? With this kind of utility, I'd think folks could user a newish Python.

### CODE REVIEW

File: compliance_checker/base.py [hunk 1]
```diff
@@ -191,9 +191,10 @@ def __del__(self):
         are cleared before the next checker uses it. Some caches were
         inadvertently mutated by other functions.
         """
-
-        cfutil.get_geophysical_variables.cache_clear()
-        cfutil.get_time_variables.cache_clear()
+        # odd errors -- module getting deleted before this object?
+        if cfutil is not None:
+            cfutil.get_geophysical_variables.cache_clear()
+            cfutil.get_time_variables.cache_clear()
```
Comment (ocefpaf) [inline compliance_checker/base.py:197]: I'd like to keep these changes in another PR so we can investigate the cause and, if this is the best solution. Let's keep the context here to just remove the last bits of pkg_resources.

File: compliance_checker/cf/util.py [hunk 2]
```diff
@@ -5,7 +5,11 @@
 
 import requests
 from cf_units import Unit
-from importlib_resources import files
+try:
+    from importlib.resources import files
+except ImportError:
+    from importlib_resources import files
```
Comment (ocefpaf) [inline compliance_checker/cf/util.py:11]: This is not necessary b/c importlib_resources is installable in Python >=3.9 and, as far as I know, it is just a stub for the std `importlib` there.

File: compliance_checker/suite.py [hunk 3]
```diff
@@ -71,10 +74,11 @@ def _get_generator_plugins(cls):
         Return a list of classes from external plugins that are used to
         generate checker classes
         """
-
+        # NOTE: updated to not use pkg_resources, but
+        #       not tested -- it is ever used?
         if not hasattr(cls, "suite_generators"):
-            gens = working_set.iter_entry_points("compliance_checker.generators")
-            cls.suite_generators = [x.resolve() for x in gens]
+            gens = impmd.entry_points(group='compliance_checker.generators')
```
Comment (ocefpaf) [inline compliance_checker/suite.py:80]:
```suggestion
            gens = importlib_metadata.entry_points(group='compliance_checker.generators')
```

File: compliance_checker/suite.py [hunk 4]
```diff
@@ -22,8 +23,10 @@
 from netCDF4 import Dataset
 from owslib.sos import SensorObservationService
 from owslib.swe.sensor.sml import SensorML
-from packaging.version import parse
-from pkg_resources import working_set
+if sys.version_info >= (3, 10):
+    import importlib.metadata as impmd
+else:
+    import importlib_metadata as impmd
```
Comment (ocefpaf) [inline compliance_checker/suite.py:29]: It is a bit clearer this 
```suggestion
import importlib_metadata
```

File: compliance_checker/suite.py [hunk 5]
```diff
@@ -136,7 +140,9 @@ def load_all_available_checkers(cls):
         Helper method to retrieve all sub checker classes derived from various
         base classes.
         """
-        cls._load_checkers(working_set.iter_entry_points("compliance_checker.suites"))
+        checkers = impmd.entry_points(group='compliance_checker.suites')
```
Comment (ocefpaf) [inline compliance_checker/suite.py:143]:
```suggestion
        checkers = importlib.metadata.entry_points(group='compliance_checker.suites')
```

File: compliance_checker/base.py (hunk 1 above)
Comment (ChrisBarker-NOAA) [reply compliance_checker/base.py:197]: Sure -- this is an odd one. I wonder if the __del__ is still needed at all.

File: compliance_checker/cf/util.py (hunk 2 above)
Comment (ChrisBarker-NOAA) [reply compliance_checker/cf/util.py:11]: This was intended as future proofing -- this way, when we drop 3.8, all we need to do is remove the dependency, rather than going in and changing all the imports later -- and it allows users with never Python versions to run without importlib_resources.

It's a matter of taste -- do with it what you will

File: compliance_checker/suite.py (hunk 3 above)
Comment (ChrisBarker-NOAA) [reply compliance_checker/suite.py:80]: See above, by giving it a name, we don't have to go in and change code when we no longer need the backport -- taste once again.

File: compliance_checker/suite.py (hunk 4 above)
Comment (ChrisBarker-NOAA) [reply compliance_checker/suite.py:29]: All the same as above -- I prefer to future-proof, and in the present not require a backport that isn't needed. But this isn't my baby.

NOTE: this one was a bit tricky -- 'cause `importlib.metadata` exists in 3.8+ (I think) but apparently the API changed more recently :-(

File: compliance_checker/cf/util.py (hunk 2 above)
Comment (ocefpaf) [reply compliance_checker/cf/util.py:11]: We would need to remove that if-clause to reduce the complexity too. In a way, `importlib_resources` is maintained and supported for py>3.9, so we don't need to change it as soon as we drop 3.8. It is a minor maintenance chore.

File: compliance_checker/suite.py (hunk 3 above)
Comment (ocefpaf) [reply compliance_checker/suite.py:80]: Compliance-checker code us huge and quite complex to read. I'd prefer to have the full import name here so we can read it and know what it is right away. I understand where you are going and I do that in my own projects. But this one is maintained by many people and I'd rather not rock the boat with style changes for minor things.

**Comment by ChrisBarker-NOAA** [comment]
I've gotten pulled away from this, and wont have any more time to look at it -- If you want to keep all the backports in for the imports, then there's only a few lines over what you already did -- probably easier to just do that on a new PR than update this one.

File: compliance_checker/base.py (hunk 1 above)
Comment (ocefpaf) [reply compliance_checker/base.py:197]: @benjwadams if you have some cycles can give us some guidance here?

**Comment by ChrisBarker-NOAA** [comment]
My thoughts:

Had we done this a couple years ago, I'd go with the backports all the way. but Python is solidly at 3.12 now -- and particularly for this kind of project (not a lib that's used in legacy code) I think focus on newer Python's makes more sense than locking in a backport that won't been needed for long.

If it was just me, I'd probably go to 3.10+ (or at least 3.9) now and be done with it :-)

At the least, I'd keep the `importlib` imports in a comment, to make the update in the future

File: compliance_checker/base.py (hunk 1 above)
Comment (ChrisBarker-NOAA) [reply compliance_checker/base.py:197]: This is really weird -- `cfutil` is a module that was imported at the top of the module. And yet it's set to None at least once when the tests are run - maybe there's a mock or something in the tests?

Sorry -- no time to dig deeper right now.

-CHB

File: compliance_checker/suite.py [hunk 6]
```diff
@@ -147,7 +153,8 @@ def _load_checkers(cls, checkers):
 
         for c in checkers:
             try:
-                check_obj = c.resolve()
+                # check_obj = c.resolve()
+                check_obj = c.load()
```
Comment (ocefpaf) [inline compliance_checker/suite.py:157]: I don't follow this change. Is it related with `pkg_resources`?

Comment (ChrisBarker-NOAA) [reply compliance_checker/suite.py:157]: yes, the pkg_resources `EntryPoint` object has a .`resolve` method, and the importlib one has `load()` instead. I"m not totally sure they are exactly the same, but this seemed to work.

File: compliance_checker/suite.py (hunk 4 above)
Comment (ocefpaf) [reply compliance_checker/suite.py:29]: Yeah. This was quite the churn and hard to follow. All these backpport libraries solved parts of the problem.

**Comment by ocefpaf** [comment]
@jcermauwedu you raised some concern on the minimum Python supported here. This PR still keeps it Ptyhon>=3.8. However, it is likely we will follow [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html) closer in the future. Would that work for your deployments?

**Comment by ocefpaf** [comment]
@benjwadams this was the last PR I wanted to get in before 5.1.1. We are good to go :shipit:

**Comment by ocefpaf** [comment]
> My thoughts:
> 
> Had we done this a couple years ago, I'd go with the backports all the way. but Python is solidly at 3.12 now -- and particularly for this kind of project (not a lib that's used in legacy code) I think focus on newer Python's makes more sense than locking in a backport that won't been needed for long.
> 
> If it was just me, I'd probably go to 3.10+ (or at least 3.9) now and be done with it :-)
> 
> At the least, I'd keep the `importlib` imports in a comment, to make the update in the future

I'd definitely go 3.10 as min but the user base here is much larger than we expected. Also, the pyupgrade pre-commit makes it super easy to fix anything that is deprecated and to adapt for new code when we change the min python, so bo need for future proofing by adding extra if-clause imports.
